internal/scan/selector.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2020 GramLabs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scan
+
+import (
+	"regexp"
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// GenericSelector can be used to implement the "Select" part of the Selector
+// interface. The "*Selector" fields are treated as Kubernetes selectors, all
+// other fields are regular expressions for matching metadata.
+type GenericSelector struct {
+	Group              string `json:"group,omitempty"`
+	Version            string `json:"version,omitempty"`
+	Kind               string `json:"kind,omitempty"`
+	Namespace          string `json:"namespace,omitempty"`
+	Name               string `json:"name,omitempty"`
+	LabelSelector      string `json:"labelSelector,omitempty"`
+	AnnotationSelector string `json:"annotationSelector,omitempty"`
+}
+
+// Select reduces the supplied resource node slice by only returning those
+// nodes which match this selector.
+func (g *GenericSelector) Select(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	m, err := newMetaMatcher(g)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]*yaml.RNode, 0, len(nodes))
+	for _, n := range nodes {
+		if meta, err := n.GetMeta(); err != nil {
+			return nil, err
+		} else if !m.matchesMeta(meta) {
+			continue
+		}
+
+		if matched, err := n.MatchesLabelSelector(g.LabelSelector); err != nil {
+			return nil, err
+		} else if !matched {
+			continue
+		}
+
+		if matched, err := n.MatchesAnnotationSelector(g.AnnotationSelector); err != nil {
+			return nil, err
+		} else if !matched {
+			continue
+		}
+
+		result = append(result, n)
+	}
+
+	return result, nil
+}
+
+// metaMatcher matches metadata. This is an alternative to `types.SelectorRegex` from Kustomize.
+type metaMatcher struct {
+	namespaceRegex *regexp.Regexp
+	nameRegex      *regexp.Regexp
+	groupRegex     *regexp.Regexp
+	versionRegex   *regexp.Regexp
+	kindRegex      *regexp.Regexp
+}
```

Comment:
why use regex instead of strings here?

as in why wouldn't we just save off `g.Namespace` ( or `strings.Split(g.Namespace, ":")[1])` ) so we could compare `string == string` in `matchesMeta`

at the end of the day I think either implementation would work, just curious.

Reply:
We allow things like `Group: "apps|extensions"`. This behavior was previously provided by Kustomize via its `types.Selector` which is used by the `ResMap` to select subsets of resources.